### PR TITLE
Game states

### PR DESCRIPTION
**Pull request by Dustfoxx**
* This is a widely encompassing update.
  * The game now has states for what stage of the game it is in.
    * PREGAME is when player are doing early setup.
    * ONGOING is when the setup is done.
    * ENDGAME is when the game is over.
    * This is an enum in the game.java file.
  * These states determine how the newturn function operates.
  * Players will now place their characters on the board before playing
  * actionHandler added to the gamecontroller.
    * Has two verisions.
    * One takes an Actions enum. This has a keyword for each possible action.
    * The other also takes a list of objects.
    * This is because "ask" needs a  feature and "move" needs to know where to move but for actions which do not require information from the can still be called without passing an object. The object list should be sent in the order in which the extra information should be passed. I.e. move in this version takes "row" and then "col" so the object being passed looks like "new Object[] {row, col}".
    * As a player has made the appropriate action for a turn (one movement and one action) newturn fires meaning "next turn" (now "end turn") is not necessary to keep the game going. This is tracked by variables within game.java.
  * checkAction has a mask creation function now for initial placing of agents.
  * AbstractCell now has a "getFootstep" function
  *  game.java now has a function to get recruits at a timestamp. This was needed as turnbar kept getting the wrong index for the turns. It now uses this function instead. 
  * as an end point a lot of these files im pretty sure are only "changed" because of some autoformatting that vscode does. Most interesting files are gameController, game.java and turnbar.
* @mrPaintMan @Vilmavds @Joakim-Johansson @Meilin77

### CODE REVIEW

**Review by mrPaintMan — COMMENTED** [review]
awesome work!!

File: Project/core/src/main/java/Controller/ActionController.java [hunk 1]
```diff
@@ -87,4 +87,32 @@ public void reveal(Footstep footstep, Board board, int[] position, List<int[]> w
...
+     * @param board        the board it is being placed on
+     * @param validityMask the mask defining valid positions. Null if all options
+     *                     are valid
+     * @param coords       coordinates the player will be placed on
+     * @return boolean defining if the action was successful or not
+     */
+    public boolean movePlayer(Player player, Board board, boolean[][] validityMask, int[] coords) {
+        // Did player choose a valid location?
+        if (validityMask != null) {
+            if (validityMask[coords[0]][coords[1]]) {
```
[the remark this answers is not in the document]
Comment (Dustfoxx) [reply Project/core/src/main/java/Controller/ActionController.java:105]: Well this is a temp function just to try out functionality, jockes function is the one we will use. And no its supposed to be true at correct positions. so there would be a "!" missing here

File: Project/core/src/main/java/Controller/CheckAction.java [hunk 2]
```diff
@@ -220,6 +220,34 @@ public boolean[][] getValidMoves(Recruiter player, Board board, int type) {
...
+    public boolean[][] getValidPlacements(Board board) {
+        int[] dims = board.getDims();
+
+        boolean[][] mask = new boolean[dims[0]][dims[1]];
+        boolean value = false;
+
+        for (int i = 0; i < dims[0]; i++) {
+            for (int j = 0; j < dims[1]; j++) {
+                value = false;
+                if (i == 0 || i == dims[1] - 1) {
```
[the remark this answers is not in the document]
Comment (Dustfoxx) [reply Project/core/src/main/java/Controller/CheckAction.java:238]: not iterate here but it should check it

**Review by Vilmavds — CHANGES_REQUESTED** [review]

**Comment by Vilmavds** [comment]
Very impressive work! 👍 🥇

**Review by Vilmavds — APPROVED** [review]

**Pull request merged**